docs/openbazaar.1
.TH OPENBAZAAR 1 "Jun 2015" "OpenBazaar 0.5.0"


.SH NAME

openbazaar \- OpenBazaar (Decentralized p2p marketplace)


.SH SYNOPSIS
openbazaar [options] <command>


.SH DESCRIPTION

\fBOpenBazaar\fR is an open source project to create a decentralized
network for commerce online (using Bitcoin) that has no fees and
cannot be censored.


.SH COMMANDS

.IP "\fBstart\fR"
Start OpenBazaar
.br
.IP "\fBstop\fR"
Stop OpenBazaar
.br
.IP "\fBhelp\fR"
Output help


.SH OPTIONS

.IP "\fB\-i\fR, \fB\-\-server\-ip <ip address>\fR"
Server public IP
.br
Notes:
  * Default value will be the external ip your network configuration exposes to the internet.
  * If '--disable-stun-check' is used and you don't specify '--server-ip' OpenBazaar
    will refuse to start unless you are on development mode ('--dev-mode').

.IP "\fB\-p\fR, \fB\-\-server\-port <port number>\fR"
Server public (P2P) port
.br
Notes:
  * Default value will be an arbitrary port number set by a STUN server check.
  * If '--disable-stun-check' is used, default value will be port 12345.
  * If you don't specify '--disable-stun-check' this number will be overwritten
    by the port number obtained via STUN check.


.IP "\fB\-k\fR, \fB\-\-http\-ip <ip address>\fR"
Web interface IP (default 127.0.0.1; use 0.0.0.0 for any)

.IP "\fB\-q\fR, \fB\-\-http\-port <port number>\fR"
Web interface port (-1 = random by default)

.IP "\fB\-l\fR, \fB\-\-log <file path>\fR"
Log file path (default 'logs/production.log')

.IP "\fB\-\-log\-level <level>\fR"
Log verbosity level (default: 30 - WARNING).
.br
Expected <level> values are:
.br
  0 - NOT SET
.br
  5 - DATADUMP
.br
  10 - DEBUG
.br
  20 - INFO
.br
  30 - WARNING
.br
  40 - ERROR
.br
  50 - CRITICAL

.IP "\fB\-d\fR, \fB\-\-dev\-mode\fR"
Enable development mode

.IP "\fB\-n\fR, \fB\-\-dev\-nodes\fR"
Number of dev nodes to start up

.IP "\fB\-\-db\-path\fR"
Database file path. (default 'db/od.db')

.IP "\fB\-\-disable\-sqlite\-crypt\fR"
Disable encryption on sqlite database

.IP "\fB\-\-bm\-user\fR"
Bitmessage API username

.IP "\fB\-\-bm\-pass\fR"
Bitmessage API password

.IP "\fB\-\-bm\-port\fR"
Bitmessage API port

.IP "\fB\-u\fR, \fB\-\-market\-id\fR"
Market ID

.IP "\fB\-j\fR, \fB\-\-disable\-upnp\fR"
Disable automatic UPnP port mappings

.IP "\fB\-\-disable\-stun\-check\fR"
Disable automatic server port setting via STUN servers (NAT Punching attempt)

.IP "\fB\-S\fR, \fB\-\-seed\-mode\fR"
Enable seed mode

.IP "\fB\-\-disable\-open\-browser\fR"
Don't open preferred web browser automatically on start

.IP "\fB\-\-config\-file\fR"
Disk path to an OpenBazaar configuration file

.IP "\fB\-\-enable\-ip\-checker\fR"
Enable periodic IP address checking.
.br
Useful in case you expect your IP to change rapidly.


.SH SEE ALSO

For more information on the usage of the program visit
<https://openbazaar.org/>

OpenBazaar home page: <https://openbazaar.org/>
.br
OpenBazaar GitHub: <https://github.com/OpenBazaar/OpenBazaar>


.SH AUTHOR

OpenBazaar is written by the OpenBazaar Team <project@openbazaar.org>
.br
This manual page was written by Angel Leon <gubatron@openbazaar.org> and Abderraouf Adjal.


.SH COPYRIGHT

Copyright (C) 2014 OpenBazaar Developers

License: OpenBazaar is licensed under the MIT license.


changelog
openbazaar (0.5.0) unstable; urgency=low

  Install/OS:
  * Install haveged package on Raspbian arch.
  * Explicitly use Python 2 on Arch distro.
  * Improve caching of pip requirements in Dockerfile.

  Documentation/UI/UX:
  * Fix endless page reload.
  * Improve rendering of long prices.
  * Fix some broken links and update README.

  Bitcoin:
  * Reject invalid BTC payment addresses.

  Network:
  * GUID is now generated from the signed BTC pubkey, to prevent
    spoofing attacks on DHT nodes (Spartacus attack).
  * Rework code for the DHT data structures (dormant for now).
  * Have peers relay only reachable close nodes.

  Bug fixes:
  * Handling of contract keywords fixed.
  * Minor other fixes.

  Miscellaneous:
  * User can specify the seeds used to bootstrap the DHT.

 -- OpenBazaar Team <project@openbazaar.org>  Wed, 3 June 2015 2:00:00 -0000

openbazaar (0.4.0) unstable; urgency=low

  Install/OS:
  * Ansible support.
  * Better install through configure.sh

  Documentation/UI/UX:
  * Animated intro gif images on main page.
  * Avatars now available for users to set in profiles.
  * Notary Services (refunds, releases)
  * Updated contract management UI (edits enabled)
  * Search cleaned up.
  * Messaging now native to OpenBazaar.

  Bitcoin:
  * HD BIP32 keys used for generating unique multi-sig addresses.

  Network:
  * Networking stack now relies on rUDP and UDP-based transport.
  * Fixes for socket overload crashes.
  * No longer using ZeroMQ for OpenBazaar. Obelisk still does.

  Testing:
  * Bandit checks enabled.
  * nosetests and jscheck enabled.

  Bug fixes:
  * Too many to count. Please refer to GitHub closed issues for this period.

  Miscellaneous:
  * No longer leverage BitMessage for internal messaging.

 -- OpenBazaar Team <project@openbazaar.org>  Tue, 21 Apr 2015 5:00:00 -0000

openbazaar (0.3.1) unstable; urgency=low
  OS/Install:
  * Increased automation of the installation script.
  * Virtualenv setup explictly requires python 2.7 on Linux and Mac.
  * Increased POSIX compliance of startup script.
  * Installed support for running on Manjaro Linux and Slackware Linux.

  Documentation/UI/UX:
  * Removed press coverage.
  * Replaced references to forums with references to the reddit and Github.
  * Updated the bitcoin donation address.
  * Other minor GUI improvements.

  Network:
  * Reduced Kademlia constants (K, cache_K), to better leverage the P2P
    network and lessen the pressure on individual nodes.
  * Increased reuse of peer connections.

  Security:
  * Disabled database encryption that was causing trouble on Macs. This is a
    temporary remedy.

  Code Quality:
  * Expanded test coverage to include the db module.
  * Simplified testing of abstract base classes.
  * Removed a few unused files and some dead code.
  * Refactored some of the logging, so that medium levels are more helpful.
  * Rewrote some code to use forward-compatible idioms and imports.
  * Cleaned up some CSS.
  * Updated dependencies.

  Bug fixes:
  * Installed synchronization for DB access.
  * Fixed a bug that could hang the stopping script.
  * Fixed argument processing in migration scripts.
  * Notaries are now properly added/removed; multiple notaries are supported.
  * Now sanitize bad URIs from other nodes that crash ZeroMQ

  Miscellaneous:
  * Restructured Makefile and test scripts to better expose the build process.
  * Installed issue handling via waffle.io.

 -- OpenBazaar Team <project@openbazaar.org>  Fri, 21 Nov 2014 12:00:00 +0000

openbazaar (0.3.0) unstable; urgency=low
  OS/Install:
  * Windows: Scripts to configure and run OpenBazaar on Windows 7 and up.
  * Mac OS X: brew upgrade is now opt-in.
  * Linux: Minor updates to debian packaging.

  Documentation/Legal:
  * Added instructions for using OpenBazaar over Tor. [Experimental]
  * Added new articles: Risk and Insurance, Notary Selection.
  * Removed residual LGPL headers.

  UI/UX:
  * Peers are now added dynamically on the sidebar.
  * Automated connecting to other peers upon node startup.
  * Replaced bash scripts for starting and stopping with a brand-new
    all-python launcher.

  Network:
  * Added support for NAT punching via STUN.
  * Changed connectivity scheme to use persistent ZMQ sockets. A node under
    normal use shouldn't crash from too many open sockets any more.

  Security:
  * Replaced sign-and-encrypt scheme with sign-then-encrypt, ensuring better
    sender anonymity.
  * Installed infrastructure for input sanitization.

  Code Quality:
  * Installed test coverage reporting via coveralls.io.
  * Added thorough unit tests for a few modules.
  * Removed a number of unused files and a lot of dead code.
  * Enforced stricter pylint checking.
  * Updated dependencies.
  * Installed support for more fine-grained logging.

  Bug fixes:
  * Ensure logs rotate properly.
  * Many minor bug fixes and stability improvements.

 -- OpenBazaar Team <project@openbazaar.org>  Sun, 09 Nov 2014 12:00:00 +0000

openbazaar (0.2.3) unstable; urgency=low
  * Messages between peers are no longer compressed.

 -- OpenBazaar Team <project@openbazaar.org>  Thu, 30 Oct 2014 15:30:00 -0500

openbazaar (0.2.2) unstable; urgency=low
  * Default logging level set to 30 -- WARNING.
  * First attempt at fixing logging issue.

 -- OpenBazaar Team <project@openbazaar.org>  Sun, 12 Oct 2014 15:30:00 -0500

openbazaar (0.2.1) unstable; urgency=low
  * Typo fix proposed by @freamon

 -- OpenBazaar Team <project@openbazaar.org>  Fri, 3 Oct 2014 15:30:00 -0500

openbazaar (0.2.0) unstable; urgency=low
  * Multiple bug fixes.
  * New feature: “Add Node” by entering Store GUID.
  * New feature: Shutdown from within the web client. (Settings > Advanced)
  * New feature: Web port is now randomized by default.
  * New feature: Automatically opens preferred web browser by default on start.
  * UX-Upgrade: stores on home page are now scrollable.
  * Tests converted to formal unit tests, travis integration.

node/openbazaar.py
#!./env/bin/python
"""
OpenBazaar launcher script.
Authors: Angel "gubatron" Leon
"""

import argparse
import multiprocessing
import os
import shlex
import sys
import threading

import psutil

import node.network_util as network_util
from node.openbazaar_daemon import node_starter, OpenBazaarContext, start_node
import node.setup_db as setup_db


def arg_to_key(arg):
    """
    Convert a long-form command line switch to an equivalent dict key,
    replacing '-' with '_'.

    Example: arg_to_key('--super-flag') == 'super_flag'
    """
    return arg.lstrip('-').replace('-', '_')


def create_argument_parser():
    defaults = OpenBazaarContext.get_defaults()
    default_db_path = os.path.join(defaults['db_dir'], defaults['db_file'])
    default_log_path = os.path.join(defaults['log_dir'], defaults['log_file'])

    parser = argparse.ArgumentParser(
        description='OpenBazaar launcher script',
        usage=usage(),
        add_help=False
    )

    # Argument entries should have the mandatory long form first.
    plain_args = (
        ('--bm-pass',),
        ('--bm-user',),
        ('--config-file',),
        ('--http-ip', '-k'),
        ('--log-level',),
        ('--market-id', '-u'),
        ('--server-ip', '-i')
    )
    for switches in plain_args:
        key = arg_to_key(switches[0])
        parser.add_argument(*switches, default=defaults[key])

    int_args = (
        ('--bm-port',),
        ('--dev-nodes', '-n'),
        ('--http-port', '-q'),
        ('--server-port', '-p'),
        ('--mediator-port',)
    )
    for switches in int_args:
        key = arg_to_key(switches[0])
        parser.add_argument(*switches, type=int, default=defaults[key])

    flags = (
        ('--dev-mode', '-d'),
        ('--disable-open-browser',),
        ('--disable-sqlite-crypt',),
        ('--disable-stun-check',),
        ('--disable-upnp', '-j'),
        ('--enable-ip-checker',),
        ('--seed-mode', '-S'),
        ('--mediator', '-m')
    )
    for switches in flags:
        key = arg_to_key(switches[0])
        parser.add_argument(
            *switches, action='store_true', default=defaults[key]
        )

    # Add miscellaneous flags.
    parser.add_argument('-s', '--seeds', nargs='*', default=defaults['seeds'])
    parser.add_argument('--db-path', default=default_db_path)
    parser.add_argument('-l', '--log', default=default_log_path)

    # Add valid commands.
    parser.add_argument('command', choices=('start', 'stop', 'help'),
                        nargs='?', default='help')

    return parser


def usage():
    return """
openbazaar [options] <command>

    COMMANDS
        start            Start OpenBazaar
        stop             Stop OpenBazaar
        help             Read this help

    EXAMPLES
        openbazaar start
        openbazaar --disable-upnp --seed-mode start
        openbazaar --enable-ip-checker start
        openbazaar -d --dev-nodes 4 -j --server-ip 79.104.98.111 start
        openbazaar --dev-mode -n 4 -i 79.104.98.111 start
        openbazaar --server-ip 200.2.8.100 --server-port 12333 --disable-stun-check start
        openbazaar --seeds seed.openbazaar.org:12345 start
        openbazaar stop

    OPTIONS
    -i, --server-ip <ip address>
        Server public IP

        Notes:
           * Default value will be the external ip your network configuration exposes to the internet.
           * If '--disable-stun-check' is used and you don't specify '--server-ip' OpenBazaar
             will refuse to start unless you are on development mode ('--dev-mode')

    -p, --server-port <port number>
        Server public (P2P) port.

        Notes:
           * Default value will be an arbitrary port number set by a STUN server check.
           * If '--disable-stun-check' is used, default value will be port 12345.
           * If you don't specify '--disable-stun-check' this number will be overwritten
             by the port number obtained via STUN check.

    -k, --http-ip <ip address>
        Web interface IP (default 127.0.0.1; use 0.0.0.0 for any)

    -q, --http-port <port number>
        Web interface port (-1 = random by default)

    -l, --log <file path>
        Log file path (default 'logs/production.log')

    --log-level <level>
        Log verbosity level (default: 30 - WARNING)
        Expected <level> values are:
           0 - NOT SET
           5 - DATADUMP
           9 - DEBUGV
          10 - DEBUG
          20 - INFO
          30 - WARNING
          40 - ERROR
          50 - CRITICAL

    -d, --dev-mode
        Enable development mode

    -n, --dev-nodes
        Number of dev nodes to start up

    --db-path
        Database file path. (default 'db/od.db')

    --disable-sqlite-crypt
        Disable encryption on sqlite database

    --bm-user
        Bitmessage API username

    --bm-pass
        Bitmessage API password

    --bm-port
        Bitmessage API port

    --mediator-port
        Act as a UDP mediator on this port

    --mediator
        Have traffic mediated through mediator server

    -u, --market-id
        Market ID

    -j, --disable-upnp
        Disable automatic UPnP port mappings

    --disable-stun-check
        Disable automatic server port setting via STUN servers (NAT Punching attempt)

    -S, --seed-mode
        Enable seed mode

    --disable-open-browser
        Don't open preferred web browser automatically on start

    --config-file
        Disk path to an OpenBazaar configuration file

    --enable-ip-checker
        Enable periodic IP address checking.
        Useful in case you expect your IP to change rapidly.

    -s, --seeds
        Specify seed servers to bootstrap the network rather than use defaults
"""


def create_openbazaar_contexts(arguments, nat_status):
    """
    Returns List<OpenBazaarContext>.

    If we are on production mode, the list will contain a
    single OpenBazaarContext object based on the arguments passed.

    If a configuration file is passed, settings from the configuration
    file will be read first, and whatever other parameters have been
    passed via the command line will override the settings on the
    configuration file.
    :rtype : list of [OpenBazaarContext]
    """
    defaults = OpenBazaarContext.get_defaults()

    server_ip = arguments.server_ip

    # "I'll purposefully leave these seemingly useless Schlemiel-styled
    # comments as visual separators to denote the beginning and end of
    # these ifs statements. They're actually nice to have when maintaining
    # so many ifs blocks. Feel free to remove post merge if they truly
    # annoy you." -Gubatron :)

    # market port
    server_port = arguments.server_port

    if nat_status is not None and not arguments.disable_stun_check:
        # unless --disable-stun-check has been passed
        # override the server ip and port for p2p communications with the ones
        # obtained from the STUN server.
        print nat_status
        server_ip = nat_status['external_ip']
        server_port = nat_status['external_port']

    # log path (requires log_dir to exist)
    if not os.path.exists(defaults['log_dir']):
        os.makedirs(defaults['log_dir'], 0o755)

    # db path
    if not os.path.exists(defaults['db_dir']):
        os.makedirs(defaults['db_dir'], 0o755)

    db_path = os.path.join(defaults['db_dir'], defaults['db_file'])
    if arguments.db_path != db_path:
        db_path = arguments.db_path

    ob_ctxs = []

    if not arguments.dev_mode:

        log_file = defaults['log_file']

        log_path = os.path.join(defaults['log_dir'], log_file)
        if arguments.log != log_path:
            log_path = arguments.log

        seed_tuples = []
        for seed in arguments.seeds:
            if not isinstance(seed, tuple):
                seed_split = seed.split(':')
                seed_tuples.append((seed_split[0], seed_split[1]))
            else:
                seed_tuples.append(seed)

        # we return a list of a single element, a production node.
        ob_ctxs.append(OpenBazaarContext(nat_status,
                                         server_ip,
                                         server_port,
                                         arguments.http_ip,
                                         arguments.http_port,
                                         db_path,
                                         log_path,
                                         arguments.log_level,
                                         arguments.market_id,
                                         arguments.bm_user,
                                         arguments.bm_pass,
                                         arguments.bm_port,
                                         arguments.mediator_port,
                                         arguments.mediator,
                                         seed_tuples,
                                         arguments.seed_mode,
                                         arguments.dev_mode,
                                         arguments.dev_nodes,
                                         arguments.disable_upnp,
                                         arguments.disable_stun_check,
                                         arguments.disable_open_browser,
                                         arguments.disable_sqlite_crypt,
                                         arguments.enable_ip_checker))
    else:
        # Create an OpenBazaarContext object for each development node.
        db_path = os.path.join(defaults['db_dir'], 'this_will_be_ignored')
        db_dirname = os.path.dirname(db_path)

        for i in range(arguments.dev_nodes):
            db_dev_filename = defaults['dev_db_file'].format(i)
            db_path = os.path.join(db_dirname, db_dev_filename)

            log_file = defaults['dev_log_file']
            dev_log_file = log_file.format(i)
            log_path = os.path.join(defaults['log_dir'], dev_log_file)

            http_port = arguments.http_port
            if http_port != 0:
                http_port += i

            if i:
                seed_mode = False
                seeds = [(server_ip, server_port)]
            else:
                seed_mode = True
                seeds = []

            ob_ctxs.append(OpenBazaarContext(nat_status,
                                             server_ip,
                                             server_port + i,
                                             arguments.http_ip,
                                             http_port,
                                             db_path,
                                             log_path,
                                             arguments.log_level,
                                             arguments.market_id,
                                             arguments.bm_user,
                                             arguments.bm_pass,
                                             arguments.bm_port,
                                             arguments.mediator_port + i,
                                             arguments.mediator,
                                             seeds,
                                             seed_mode,
                                             arguments.dev_mode,
                                             arguments.dev_nodes,
                                             arguments.disable_upnp,
                                             arguments.disable_stun_check,
                                             arguments.disable_open_browser,
                                             arguments.disable_sqlite_crypt,
                                             arguments.enable_ip_checker))
    return ob_ctxs


def ensure_database_setup(ob_ctx, defaults):
    db_path = ob_ctx.db_path
    default_db_path = os.path.join(defaults['db_dir'], defaults['db_file'])
    default_dev_db_path = os.path.join(defaults['db_dir'],
                                       defaults['dev_db_file'])

    if ob_ctx.dev_mode and db_path == default_db_path:
        # override default db_path to developer database path.
        db_path = default_dev_db_path

    # make sure the folder exists wherever it is
    db_dirname = os.path.dirname(db_path)
    if not os.path.exists(db_dirname):
        os.makedirs(db_dirname, 0o755)

    if not os.path.exists(db_path):
        # setup the database if file not there.
        print "[openbazaar] bootstrapping database ", os.path.basename(db_path)
        setup_db.setup_db(db_path, ob_ctx.disable_sqlite_crypt)
        print "[openbazaar] database setup completed\n"


def start(arguments):
    defaults = OpenBazaarContext.get_defaults()
    network_util.set_stun_servers()

    # Turn off checks that don't make sense in development mode
    if arguments.dev_mode:
        print "DEVELOPMENT MODE! (Disable STUN check and UPnP mappings)"
        arguments.disable_stun_check = True
        arguments.disable_upnp = True

    # Try to get NAT escape UDP port
    if not arguments.dev_mode:
        nat_status = network_util.get_nat_status()
    else:
        nat_status = {
            'nat_type': 'Restric NAT'
        }

    if not arguments.disable_stun_check:
        print "Checking NAT Status..."
        if nat_status.get('nat_type') == 'Blocked':
            print "openbazaar: Could not start. The network you are on currently blocks usage",
            print "of OpenBazaar."
            print "(We currently do not support usage on completely closed networks.)"
            sys.exit(1)
    elif not arguments.dev_mode and network_util.is_private_ip_address(arguments.server_ip):
        print "openbazaar: Could not start. The given/default server IP address",
        print arguments.server_ip, "is not a public ip address."
        print "(Try './openbazaar help' and read about the '--server-ip', '-i' options)"
        sys.exit(1)

    ob_ctxs = create_openbazaar_contexts(arguments, nat_status)

    for ob_ctx in ob_ctxs:
        ensure_database_setup(ob_ctx, defaults)

    if hasattr(sys, 'frozen'):
        start_node(ob_ctxs[0])
    else:
        process = multiprocessing.Process(target=node_starter,
                                          args=(ob_ctxs,))
        process.start()


def terminate_or_kill_process(process):
    try:
        process.terminate()  # in POSIX, sends SIGTERM.
        process.wait(5)
    except psutil.TimeoutExpired:
        try:
            print "process {0} didn't terminate - sending sigkill".format(process.pid)
            process.kill()  # sends KILL signal.
            process.wait(5)
        except psutil.TimeoutExpired:
            print 'timeout waiting for process {0} to end'.format(process.pid)
            return


def stop():
    my_pid = os.getpid()  # don't kill the killer.
    for process in psutil.process_iter():
        try:
            pdict = process.as_dict()
            if my_pid != int(pdict['pid']) and pdict['cmdline'] is not None:
                cmd = ' '.join(pdict['cmdline'])
                if cmd.find('openbazaar') > -1 and cmd.find('start') > -1:
                    threading.Thread(
                        target=terminate_or_kill_process,
                        args=(process,)
                    ).start()
        except psutil.NoSuchProcess:
            pass


def load_config_file_arguments(parser):
    """
    Load configuration file into sys.argv for argument parsing. Insert
    config file arguments before command line arguments so that command
    line arguments have higher precedence.
    """
    parsed_arguments = parser.parse_args()
    if parsed_arguments.config_file is not None:
        try:
            with open(parsed_arguments.config_file) as opened_file:
                sys.argv[1:1] = shlex.split(opened_file.read(), comments=True)
        except IOError as err:
            print "NOTICE: Ignoring invalid config file: ",
            print parsed_arguments.config_file
            print err


def main():
    parser = create_argument_parser()
    load_config_file_arguments(parser)
    arguments = parser.parse_args()

    if arguments.command == 'start':
        start(arguments)
    elif arguments.command == 'stop':
        stop()
    elif arguments.command == 'help':
        print usage()


if __name__ == '__main__':
    main()
